### ISSUE TITLE
Mapped property for one member of result union not supported

### ISSUE THREAD

**Issue by xirzec** [Azure/autorest.typescript#1921]
We have an interesting case in OpenAI where one side of a result type had a mapped property, so the generated deserialization function didn't compile:

```ts
export async function _getAzureBatchImageGenerationOperationStatusDeserialize(
  result:
    | GetAzureBatchImageGenerationOperationStatus200Response
    | GetAzureBatchImageGenerationOperationStatusDefaultResponse
): Promise<BatchImageGenerationOperationResponse> {
  if (isUnexpected(result)) {
    throw result.body;
  }

  return {
    id: result.body["id"],
    created: result.body["created"],
    expires: result.body["expires"],
    result: !result.body.result
      ? undefined
      : {
          created: result.body.result?.["created"],
          data: result.body.result?.["data"],
        },
    status: result.body["status"],
    error: !result.body.error ? undefined : result.body.error,
  };
}
```

The problem is that in the case that `result.body.result.data` is a `ImagePayload[]`, the wire property of `b64_json` needs to be converted to the friendly name of `base64Data`.

We worked around this for now by writing a manual conversion helper:

```ts
/**
 * Convert REST-level ImageGenerationsOutput.data to HLC-level ImageGenerations.data
 * @internal
 */
function toImageGenerationsData(
  input: ImageGenerationsOutput["data"]
): ImageLocation[] | ImagePayload[] {
  return input.map(function (locationOrPayload) {
    if ("url" in locationOrPayload) {
      return locationOrPayload as ImageLocation;
    } else {
      return {
        base64Data: locationOrPayload.b64_json,
      } as ImagePayload;
    }
  }) as ImageLocation[] | ImagePayload[];
}
```

This is a tricky problem to resolve since I think if the result was just one of these types we could unwind it and map it in a literal, but since it's a union we have to somehow generate code that only applies to the case where the property to be mapped exists.

**Comment by qiaozha** [comment]
Hi @xirzec,  I found they are using projectedName json in their [typespec](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/OpenAI.Inference/models/images.tsp#L69-L73)
```
@doc("An image response item that directly represents the image data as a base64-encoded string.")
@added(ServiceApiVersions.v2023_06_01_Preview)
model ImagePayload {
  @doc("The complete data for an image represented as a base64-encoded string.")
  @projectedName("json", "b64_json")
  base64Data: string;
}
```
I kind of don't get it. Is the property name of ImagePayload `base64Data` or `b64_json` in the rest api layer? if it's `base64Data` we probably should remove this projectedName json?  If it's b64_json, then RLC should keep consistent with the rest api layer? Or is there anything special about base64 encoding I should know ? Thanks :)

**Comment by qiaozha** [comment]
@xirzec does type predicates look better in this case ? We can't use overload for this, because it will requires us to give two unique model names from the top of the response type `BatchImageGenerationOperationResponse`

```typescript
interface ImageLocation {
  url: string
}
interface ImagePayload {
  base64Data: string;
}

interface ImageRestPayload {
  b64_json: string
}

function convertRestPayloadToModularPayload(data: ImageLocation[] | ImageRestPayload[]): ImageLocation[] | ImagePayload[] {
  if (isImageLocation(data)) {
    return data;
  }
  return convertImageRestPayloadToImagePayload(data);
}

function isImageLocation(item: ImageLocation[] | ImageRestPayload[]): item is ImageLocation[] {
  if (item.length > 0) {
    return (item as ImageLocation[])[0].url !== undefined;
  }
  return false;
}

function convertImageRestPayloadToImagePayload(data: ImageRestPayload[]): ImagePayload[] {
  return data.map(item => {
    return {
      base64Data: item.b64_json
    }
  })
}
```
I kind of prefer this way, because it might be easier for codegen if there's a lot of common properties between ImageLocation and ImageRestPayload and if there's another type added into the union array, we could just generate one more type predicates function.

**Comment by xirzec** [comment]
@qiaozha I think the above is a fine way for codegen to handle this kind of scenario. I agree codegen shouldn't be in the business of changing API patterns, so either the service could change to use an overload in TSP if they wanted an overload or the package owner could decide to handle that in the customization layer.

**Comment by qiaozha** [comment]
@xirzec Currently there're some work that are trying to summarize all kinds of valid scenarios in tsp Union usages. I will get back to this to see if we need revisit the proposal and then work on it. Let me know if it works for you? Thanks

**Comment by qiaozha** [comment]
@xirzec As I found that we are treating union as anyOf in TypeSpec and it's been pointed out by Brian that making it oneOf would require us to do extra logic to validate this object does not belong to other union variants,  Also, as being pointed out by Will here  https://github.com/microsoft/typespec/issues/2378#issuecomment-1712086924 it will might be incorrect to speculate the type just based on the required properties. then I realize we don't need to speculate which union variant those properties belong to https://github.com/microsoft/typespec/issues/2378#issuecomment-1712424281. So we could just change the above function of `convertRestPayloadToModularPayload` like below.  Let me know how you think ? 

```
function convertRestPayloadToModularPayload(data: ImageLocation[] | ImageRestPayload[]): ImageLocation[] | ImagePayload[] {
  return data.map(item => {
    return {
      base64Data: (item as ImageRestPayload)['b64_json'],
      url: (item as ImageLocation)['url']
    }
  })
}
```

**Comment by xirzec** [comment]
@qiaozha I like how much simpler that codegen is. 

In this case it works out to assume `map` since each variant is an Array, though we might have to do a small test like `Array.isArray()` if one variant was not an array type.

**Comment by qiaozha** [comment]
Yeah, we will probably need to add a filter before, I haven't thought through the details yet.

**Comment by MaryGao** [comment]
TL;DR: I don't think we could have a general sulotion to cover all cases to deserialize heterogeneous data without runtime business data so I'd like to confirm - what kind of cases would be supported and why? if not support, how to generate them?

**Here is the whole story:**
 
I think the above the unioned properties could work because `ImageLocation` and `ImageRestPayload` don't have overlapped properties.

Things will be different if they have some common properties like below cases, without knowing business details we can't  speculate which properties in RLC layer should be used for Modular `a` property.
```typescript
// Modular models
interface Model {
  prop: Foo | Bar; 
}

interface Foo {
  a: string;
}

interface Bar {
  a: Date;
}

// RLC models
interface FooA {
  aName: string;
}
interface BarA {
  aNameAlias: string;
}
function convertUnion(data: FooA | BarA): Foo| Bar {
return {
  a: data.aName // ===> How to know we should map aName to string a OR aNameAlias to Date a?
  // ...
}
}
``` 

Taking the union as anyOf in TypeScript is acceptable for me and this is somehow the current type system work, you could see [the demo](https://www.typescriptlang.org/play?#code/C4TwDgpgBACg9gSwHbCgXigbwFBT1ADwC4okBXAWwCMIAnAbl3xBPOrsYF9HtRIoAMgEMaAG3RYmeJEIoQSAZ2C1kAcy48AxnCRKowABZqS8ZKgA+gkRHEYc+QiQAMAGilQWUV+5lySAImAIJX8oAHowqDIDAFo4AwBCbG4gA). But we should be careful with [discriminating union](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) because the type system could narrow down the type correctly so it may take it as oneOf ([example here](https://www.typescriptlang.org/play?#code/PTAEBUE8AcFMGUDGAnAltALgKAzWoAZAewEMATVAOwHN4MSN8BeUAby1FAGd7GAuUACIANqQo1BAblAhQ9algC+knHlAAxEqmGwydBszYduvWAMEAzLTrJSZYecZ4MArlwGUXAWwBGsZCrKqnCg8C6IiLBcXPqMoCzsnM78QlzhkdF2so6cyFHQRJRcZkacnKhcxOS6Aj5ERDoklCqcQUFYiIU8oCSgAmERUTGmoAA+hGJUtCMJTqbmouRTggA0xnlcBUUliWUVVWQ1csgusMaKSkA)).

No matter the unioned property or speculating the type there will be situations where we can't handle. So I'd like to have a clear answer with below questions:
- Do we need support union? 
[Mary] It would be nice to have this supported because we could easily define union in TypeSpec. But I would take this as a low confidence feature, other languages only provide protocol method if contains unions & I would assume this feature may be not widely used.

- What kind of union cases should be supported and why?
[Mary] It would be hard to make a proper assumption because we didn't receive any union cases from Swagger and only having OpenAI case from TypeSpec. If we could have list of cases which could help us design better.

- How to handle un-supported cases?
[Mary] I may ask another question - for un-supported cases, do we always require customizations from the persons preparing SDK? Things would be simpler if the ones could mitigate the gaps of emitter.   For modular we could 1) either not generate this operation like other languages or 2) take this `as any` and skip the type checking and deserilization process. The concern for the former one is we plan to spilt the RLC and modular, end-users have to import another library and to touch this operation. For the latter one the returned type actually mix the RLC and Modular formats, it would be okay if we have customizations but it would be quite wired if we directly expose these interfaces to end users.

**Comment by xirzec** [comment]
This is a great question! For RLC I believe we sidestep this problem since we don't offer rich deserialization and it's up to the consumer to tease apart a union type with whatever they need to discriminate it.

For Modular, since we are trying to support scenarios like coercing to Date objects and renaming nested properties, we have to be a little more thoughtful.

Here's a proposal I think might work for us:

1. If an [`@discriminator`](https://microsoft.github.io/typespec/standard-library/built-in-decorators#@discriminator) decorator exists for a union, use that to generate a simple typeguard helper
2. If no `@discriminator` is present (like in OpenAI), generate a typeguard helper that throws a "not implemented" exception

In the second case, it will be required that the developer building the SDK needs to use customization to implement the guard correctly for the library to function.

This allows for a happy path where the spec author is considerate enough to provide a way to easily discriminate the union by checking a single property while allowing for manual intervention when this consideration is not given.

Thoughts?

**Comment by lirenhe** [comment]
I like this idea, and it is similar to the .NET team's effort to handle those cases.

And I suggest we should also consider:
1. The grow up story for service team to implement this function to remove ' "not implemented" exception' and return the correct result.
2. How we could leverage client.tsp so that with some additional information provided, we could auto generate that function correctly. (without "not implemented" exception)
